File: backend/python-api/app/service/twofa_service.py
from __future__ import annotations

import asyncio
import logging
import secrets
from datetime import datetime, timedelta, timezone

import pyotp
from passlib.hash import bcrypt

from app.domain.errors import ERR_INVALID_INPUT, ERR_NOT_FOUND, ERR_UNAUTHORIZED, DomainError
from app.domain.login_event import LoginEvent
from app.domain.twofa import RecoveryCode, TwoFactorToken
from app.dto.auth import TokenPair
from app.platform.email import EmailSender
from app.repository.login_event_repo import LoginEventRepository
from app.repository.twofa_repo import TwoFARepository
from app.repository.user_repo import UserRepository
from app.service.token_service import TokenService

logger = logging.getLogger(__name__)

MAX_2FA_ATTEMPTS = 5
TWO_FA_EXPIRY_MINUTES = 5
RECOVERY_CODE_COUNT = 10


class TwoFAService:
    def __init__(
        self,
        users: UserRepository,
        twofa_repo: TwoFARepository,
        login_events: LoginEventRepository,
        token_svc: TokenService,
        email_sender: EmailSender,
    ):
        self._users = users
        self._twofa = twofa_repo
        self._login_events = login_events
        self._token_svc = token_svc
        self._email = email_sender

    async def send_2fa_code(self, user_id: int, remember_me: bool, session) -> None:
        user = await self._users.find_by_id(user_id)
        await self._twofa.invalidate_tokens(user.id)

        code = f"{secrets.randbelow(900000) + 100000}"
        token = TwoFactorToken(
            user_id=user.id,
            code=code,
            remember_me=remember_me,
            expires_at=datetime.now(timezone.utc) + timedelta(minutes=TWO_FA_EXPIRY_MINUTES),
        )
        await self._twofa.create_token(token)
        await session.commit()

        asyncio.create_task(
            self._email.send_2fa_code(user.email, code, user.language or "en", user.full_name)
        )

    async def verify_2fa(
        self, public_id: str, code: str, client_ip: str, user_agent: str, session
    ) -> tuple[User, TokenPair]:
        user = await self._users.find_by_public_id(public_id)
        token = await self._twofa.find_latest_unused_token(user.id)

        if token is None:
            raise DomainError(ERR_INVALID_INPUT, "No active 2FA code found")

        if token.attempts >= MAX_2FA_ATTEMPTS:
            await self._twofa.invalidate_tokens(user.id)
            await session.commit()
            raise DomainError(ERR_UNAUTHORIZED, "Too many failed attempts")

        if datetime.now(timezone.utc) > token.expires_at.replace(tzinfo=timezone.utc):
            raise DomainError(ERR_UNAUTHORIZED, "Verification code has expired")

        if token.code != code:
            await self._twofa.increment_attempts(token.id)
            await session.commit()
            raise DomainError(ERR_UNAUTHORIZED, "Invalid verification code")

        await self._twofa.mark_token_used(token.id)
        pair = await self._token_svc.issue_token_pair(user.id, user.role, token.remember_me)
        event = LoginEvent(user_id=user.id, ip=client_ip, user_agent=user_agent)
        await self._login_events.create(event)
        await session.commit()

        return user, pair

    async def resend_2fa(self, public_id: str, session) -> None:
        try:
            user = await self._users.find_by_public_id(public_id)
        except DomainError:
            return

        if user.two_fa_secret:
            raise DomainError(ERR_INVALID_INPUT, "TOTP is enabled. Use your authenticator app.")

        await self._twofa.invalidate_tokens(user.id)
        code = f"{secrets.randbelow(900000) + 100000}"
        token = TwoFactorToken(
            user_id=user.id,
            code=code,
            expires_at=datetime.now(timezone.utc) + timedelta(minutes=TWO_FA_EXPIRY_MINUTES),
        )
        await self._twofa.create_token(token)
        await session.commit()

        asyncio.create_task(
            self._email.send_2fa_code(user.email, code, user.language or "en", user.full_name)
        )

    async def setup_2fa(self, user_id: int, session) -> dict:
        user = await self._users.find_by_id(user_id)
        secret = pyotp.random_base32()
        totp = pyotp.TOTP(secret)
        uri = totp.provisioning_uri(name=user.email, issuer_name="SaaS Starter")

        user.two_fa_secret = secret
        await self._users.update(user)
        await session.commit()

        return {"otpauth_url": uri, "secret": secret, "message": "Scan the QR code with your authenticator app"}

    async def verify_2fa_setup(self, user_id: int, code: str, session) -> dict:
        user = await self._users.find_by_id(user_id)

        if not user.two_fa_secret:
            raise DomainError(ERR_INVALID_INPUT, "2FA has not been set up")

        totp = pyotp.TOTP(user.two_fa_secret)
        if not totp.verify(code):
            raise DomainError(ERR_UNAUTHORIZED, "Invalid TOTP code")

        user.two_fa_enabled = True
        await self._users.update(user)

        recovery_codes: list[str] = []
        for _ in range(RECOVERY_CODE_COUNT):
            raw = secrets.token_hex(4)
            recovery_codes.append(raw)
            rc = RecoveryCode(
                user_id=user.id,
                code_hash=bcrypt.using(rounds=12).hash(raw),
            )
            await self._twofa.create_recovery_code(rc)

        await session.commit()

        return {
            "message": "Two-factor authentication enabled successfully",
            "recovery_codes": recovery_codes,
            "warning": "Save these recovery codes in a safe place. Each code can only be used once.",
        }

    async def verify_totp_login(
        self, public_id: str, code: str, remember_me: bool, client_ip: str, user_agent: str, session
    ) -> tuple[User, TokenPair]:
        user = await self._users.find_by_public_id(public_id)

        if not user.two_fa_secret:
            raise DomainError(ERR_INVALID_INPUT, "TOTP is not set up")

        totp = pyotp.TOTP(user.two_fa_secret)
        if not totp.verify(code):
            raise DomainError(ERR_UNAUTHORIZED, "Invalid TOTP code")

        pair = await self._token_svc.issue_token_pair(user.id, user.role, remember_me)
        event = LoginEvent(user_id=user.id, ip=client_ip, user_agent=user_agent)
        await self._login_events.create(event)
        await session.commit()

        return user, pair

    async def use_recovery_code(
        self, public_id: str, code: str, remember_me: bool, client_ip: str, user_agent: str, session
    ) -> tuple[User, TokenPair]:
        user = await self._users.find_by_public_id(public_id)
        unused_codes = await self._twofa.find_unused_recovery_codes(user.id)

        matched: RecoveryCode | None = None
        for rc in unused_codes:
            if bcrypt.verify(code, rc.code_hash):
                matched = rc
                break

        if matched is None:
            raise DomainError(ERR_UNAUTHORIZED, "Invalid or already used recovery code")

        await self._twofa.mark_recovery_code_used(matched.id)
        pair = await self._token_svc.issue_token_pair(user.id, user.role, remember_me)
        event = LoginEvent(user_id=user.id, ip=client_ip, user_agent=user_agent)
        await self._login_events.create(event)
        await session.commit()

        return user, pair

File: backend/go-api/internal/service/twofa_service.go
package service

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"log"
	"time"

	"saas-starter/backend/go-api/internal/domain"
	"saas-starter/backend/go-api/internal/dto"
	"saas-starter/backend/go-api/internal/platform/email"
	"saas-starter/backend/go-api/internal/repository"

	"github.com/pquerna/otp/totp"
	"golang.org/x/crypto/bcrypt"
)

type TwoFAService struct {
	users     repository.UserRepository
	twofa     repository.TwoFARepository
	txManager repository.TxManager
	tokenSvc  *TokenService
	email     email.Sender
	events    repository.LoginEventRepository
}

func NewTwoFAService(
	users repository.UserRepository,
	twofa repository.TwoFARepository,
	txManager repository.TxManager,
	tokenSvc *TokenService,
	emailSender email.Sender,
	events repository.LoginEventRepository,
) *TwoFAService {
	return &TwoFAService{
		users: users, twofa: twofa, txManager: txManager,
		tokenSvc: tokenSvc, email: emailSender, events: events,
	}
}

func (s *TwoFAService) Verify2FA(ctx context.Context, req dto.Verify2FARequest, clientIP, userAgent string) (*LoginResult, error) {
	user, err := s.users.FindByPublicID(ctx, req.UserID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NewError(domain.ErrUnauthorized, "invalid 2FA request")
		}
		return nil, err
	}

	latest, err := s.twofa.FindLatestUnusedToken(ctx, user.ID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NewError(domain.ErrUnauthorized, "no pending 2FA code")
		}
		return nil, err
	}

	if latest.Attempts >= domain.MaxTwoFAAttempts {
		_ = s.twofa.InvalidateTokens(ctx, user.ID)
		return nil, domain.NewError(domain.ErrTooManyAttempts, "too many failed attempts, request a new code")
	}

	if time.Now().After(latest.ExpiresAt) {
		return nil, domain.NewError(domain.ErrTokenExpired, "2FA code expired")
	}

	if latest.Code != req.Code {
		_ = s.twofa.IncrementAttempts(ctx, latest.ID)
		return nil, domain.NewError(domain.ErrUnauthorized, "invalid 2FA code")
	}

	if err := s.twofa.MarkTokenUsed(ctx, latest.ID); err != nil {
		return nil, err
	}

	pair, err := s.tokenSvc.IssueTokenPair(ctx, user.ID, user.Role, latest.RememberMe)
	if err != nil {
		return nil, err
	}

	go s.recordLogin(user.ID, clientIP, userAgent)

	return &LoginResult{User: user, TokenPair: pair}, nil
}

func (s *TwoFAService) Resend2FA(ctx context.Context, publicID string) error {
	user, err := s.users.FindByPublicID(ctx, publicID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil
		}
		return err
	}

	if user.TwoFAEnabled && user.TwoFASecret != "" {
		return domain.NewError(domain.ErrInvalidInput, "TOTP is enabled; use authenticator app")
	}

	s.twofa.InvalidateTokens(ctx, user.ID)

	code := generateSecure6DigitCode()
	token := &domain.TwoFactorToken{
		UserID:    user.ID,
		Code:      code,
		ExpiresAt: time.Now().Add(5 * time.Minute),
	}
	if err := s.twofa.CreateToken(ctx, token); err != nil {
		return err
	}

	go func() {
		if err := s.email.Send2FACode(user.Email, code, user.PreferredLang("en"), user.FullName()); err != nil {
			log.Printf("Failed to send 2FA code to %s: %v", user.Email, err)
		}
	}()

	return nil
}

func (s *TwoFAService) Setup2FA(ctx context.Context, userID uint) (*dto.Setup2FAResponse, error) {
	user, err := s.users.FindByID(ctx, userID)
	if err != nil {
		return nil, err
	}

	key, err := totp.Generate(totp.GenerateOpts{
		Issuer:      "SaaS Starter",
		AccountName: user.Email,
	})
	if err != nil {
		return nil, err
	}

	user.TwoFASecret = key.Secret()
	if err := s.users.Update(ctx, user); err != nil {
		return nil, err
	}

	return &dto.Setup2FAResponse{
		OTPAuthURL: key.URL(),
		Secret:     key.Secret(),
	}, nil
}

func (s *TwoFAService) Verify2FASetup(ctx context.Context, userID uint, code string) (*dto.Verify2FASetupResponse, error) {
	user, err := s.users.FindByID(ctx, userID)
	if err != nil {
		return nil, err
	}

	if user.TwoFASecret == "" {
		return nil, domain.NewError(domain.ErrInvalidInput, "2FA not set up")
	}

	if !totp.Validate(code, user.TwoFASecret) {
		return nil, domain.NewError(domain.ErrUnauthorized, "invalid 2FA code")
	}

	var recoveryCodes []string
	err = s.txManager.WithTx(ctx, func(txCtx context.Context) error {
		user.TwoFAEnabled = true
		if err := s.users.Update(txCtx, user); err != nil {
			return err
		}

		for i := 0; i < 10; i++ {
			b := make([]byte, 8)
			if _, err := rand.Read(b); err != nil {
				return err
			}
			plainCode := hex.EncodeToString(b)
			recoveryCodes = append(recoveryCodes, plainCode)

			hashed, err := bcrypt.GenerateFromPassword([]byte(plainCode), BcryptCost)
			if err != nil {
				return err
			}
			rc := &domain.RecoveryCode{UserID: user.ID, CodeHash: string(hashed)}
			if err := s.twofa.CreateRecoveryCode(txCtx, rc); err != nil {
				return err
			}
		}
		return nil
	})
	if err != nil {
		return nil, err
	}

	return &dto.Verify2FASetupResponse{
		RecoveryCodes: recoveryCodes,
	}, nil
}

func (s *TwoFAService) VerifyTOTPLogin(ctx context.Context, req dto.VerifyTOTPLoginRequest, clientIP, userAgent string) (*LoginResult, error) {
	user, err := s.users.FindByPublicID(ctx, req.UserID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NewError(domain.ErrUnauthorized, "invalid request")
		}
		return nil, err
	}

	if !user.TwoFAEnabled || user.TwoFASecret == "" {
		return nil, domain.NewError(domain.ErrInvalidInput, "TOTP not enabled")
	}

	if !totp.Validate(req.Code, user.TwoFASecret) {
		return nil, domain.NewError(domain.ErrUnauthorized, "invalid 2FA code")
	}

	pair, err := s.tokenSvc.IssueTokenPair(ctx, user.ID, user.Role, req.RememberMe)
	if err != nil {
		return nil, err
	}

	go s.recordLogin(user.ID, clientIP, userAgent)

	return &LoginResult{User: user, TokenPair: pair}, nil
}

func (s *TwoFAService) UseRecoveryCode(ctx context.Context, req dto.UseRecoveryCodeRequest, clientIP, userAgent string) (*LoginResult, error) {
	user, err := s.users.FindByPublicID(ctx, req.UserID)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NewError(domain.ErrUnauthorized, "invalid request")
		}
		return nil, err
	}

	codes, err := s.twofa.FindUnusedRecoveryCodes(ctx, user.ID)
	if err != nil {
		return nil, err
	}

	var matchedCode *domain.RecoveryCode
	for i := range codes {
		if bcrypt.CompareHashAndPassword([]byte(codes[i].CodeHash), []byte(req.Code)) == nil {
			matchedCode = &codes[i]
			break
		}
	}

	if matchedCode == nil {
		return nil, domain.NewError(domain.ErrUnauthorized, "invalid recovery code")
	}

	if err := s.twofa.MarkRecoveryCodeUsed(ctx, matchedCode.ID); err != nil {
		return nil, err
	}

	pair, err := s.tokenSvc.IssueTokenPair(ctx, user.ID, user.Role, req.RememberMe)
	if err != nil {
		return nil, err
	}

	go s.recordLogin(user.ID, clientIP, userAgent)

	return &LoginResult{User: user, TokenPair: pair}, nil
}

func (s *TwoFAService) recordLogin(userID uint, ip, userAgent string) {
	event := &domain.LoginEvent{
		UserID: userID, IP: ip, UserAgent: userAgent, LoggedAt: time.Now(),
	}
	if err := s.events.Create(context.Background(), event); err != nil {
		log.Printf("Failed to record login event for user %d: %v", userID, err)
	}
}

func generateSecure6DigitCode() string {
	b := make([]byte, 3)
	rand.Read(b)
	num := (int(b[0])<<16 | int(b[1])<<8 | int(b[2])) % 1000000
	return fmt.Sprintf("%06d", num)
}
